### PR TITLE
[CoSim] enable solvers to run in serial/less MPI-procs

### PR DESCRIPTION
**Pull request by philbucher**
This is my final masterpiece :smiley: 

Now it is finally possible to run Kratos in MPI, but the individual solvers use less MPI-processes or even run in serial!
E.g. for FSI it is now possible to run the structure in serial and the Fluid in MPI. Especially helpful with complex structures where the solvers that we have available in MPI are not working well.
Also now the SDof solvers can be run only on one rank (FYI @mpentek @mtnzguillermo)

By default nothing changes, i.e. unless implemented then the solvers will use all MPI-processes
I modified the SDof solver-wrapper so that the sdof solver runs only on one rank.
Furthermore I adapted the Kratos-fluid and the Kratos-structure wrapper to support the new feature.

The mapper was the biggest blocker, it was already adapted in #9182

image

@KlausBSautter @VeronikaSinger theoretically now you could run super expensive structural simulations in MPI and the particles in serial as before. Let me know if you are interested :)

prerequisite for #9166

### CODE REVIEW

File: applications/CoSimulationApplication/python_scripts/data_transfer_operators/copy_single_to_distributed.py [hunk 1]
```diff
@@ -9,26 +9,41 @@ class CopySingleToDistributed(CoSimulationDataTransferOperator):
...
+
+        if from_solver_data.IsDefinedOnThisRank():
+            data_value = from_solver_data.GetData()
+
+            if from_solver_data.IsDistributed():
+                raise Exception("The source if the data cannot be distributed!")
 
-        if not data_value.size == 1:
-            raise Exception('Interface data "{}" of solver "{}" requires to be of size 1, got: {}'.format(from_solver_data.name, from_solver_data.solver_name, data_value.size))
+            if not data_value.size == 1:
```
Comment (pooyan-dadvand) [inline applications/CoSimulationApplication/python_scripts/data_transfer_operators/copy_single_to_distributed.py:21]: Why this restriction?

File: applications/CoSimulationApplication/python_scripts/data_transfer_operators/sum_distributed_to_single.py [hunk 2]
```diff
@@ -12,17 +12,25 @@ class SumDistributedToSingle(CoSimulationDataTransferOperator):
...
-        if from_solver_data.IsDistributed():
-            value = from_solver_data.GetModelPart().GetCommunicator().GetDataCommunicator().SumAll(value)
+        value = from_solver_data.GetModelPart().GetCommunicator().GetDataCommunicator().Sum(value, 0)
         summed_data_array = np.array([value])
 
+        if not to_solver_data.IsDefinedOnThisRank():
+            return
+
+        if to_solver_data.IsDistributed():
+            raise Exception("The destination if the data cannot be distributed!")
```
Comment (pooyan-dadvand) [inline applications/CoSimulationApplication/python_scripts/data_transfer_operators/sum_distributed_to_single.py:28]: typo

Comment (pooyan-dadvand) [reply applications/CoSimulationApplication/python_scripts/data_transfer_operators/sum_distributed_to_single.py:28]: Now, I am not sure if I am in the same page....

I thought that the transfer should work via mapper if you have some to some transfer. Isn't it?

**Review by pooyan-dadvand — COMMENTED** [review]
Nice work! Some small comments

File: applications/CoSimulationApplication/python_scripts/data_transfer_operators/sum_distributed_to_single.py (hunk 2 above)
Comment (philbucher) [reply applications/CoSimulationApplication/python_scripts/data_transfer_operators/sum_distributed_to_single.py:28]: this has nothing to do with the mapper, this is the data-transfer-operator for sof solvers
=> both the `KratosMappingDataTransferOperator` and the `SumDistributedToSingleDataTransferOperator` derive from the baseclass `DataTransferOperator`

In general the `DataTransferOperator` is in charge of exchanging data between discretizations
If both of them are meshes, then you use mapping
if one of them is an sdof (which we use for bridge flutter simulation) then you would use this one

Bottom line: Not all solvers use meshes and hence mapping is not always needed.

**Review by pooyan-dadvand — APPROVED** [review]
👍
